backend/module/audit/audit.go

```diff
@@ -75,12 +76,49 @@ func (m *mod) GetEvents(ctx context.Context, req *auditv1.GetEventsRequest) (*au
 		if err != nil {
 			return nil, err
 		}
-		return &auditv1.GetEventsResponse{Events: events}, nil
+
+		resp.Events = events
 	default:
 		return nil, errors.New("no time window requested")
 	}
+
+	if req.PageToken != "" {
+		page, err := strconv.Atoi(req.PageToken)
+		if err != nil {
+			return nil, fmt.Errorf("invalid page token: %s", req.PageToken)
+		}
+		limit := int(req.Limit)
+		if req.Limit == 0 {
+			limit = 10
+		}
+		offset := page * limit
+		if offset > len(resp.Events) {
+			offset = 0
+		}
+
+		end := offset + limit
+		if end < len(resp.Events) {
+			resp.NextPageToken = strconv.FormatInt(int64(page+1), 10)
+		}
+
+		if end > len(resp.Events) {
+			end = len(resp.Events)
+		}
+
+		resp.Events = resp.Events[offset:end]
```

Comment:
should we try to push this to the query layer instead?

Reply:
yeah, we could but I was thinking that would be a breaking change since the query would need to take the start and stop props. Maybe I can do it in a separate PR, what do you think?

Reply:
i'd go ahead and do it now, an audit query could take down the DB or service otherwise with the potential for the amount of data it could return

Reply:
couldn't they currently take it down now if the range is large enough?